redaxo/src/addons/structure/lib/navigation.php

```diff
@@ -330,32 +351,103 @@ protected function _getNavigation($category_id, $depth = 1)
                 ) {
-                    $l .= $this->_getNavigation($nav->getId(), $depth);
+                    $link .= PHP_EOL.$this->_getNavigation($nav->getId(), $depth);
                 }
                 --$depth;
-                $l .= '</li>';
-                $lis[] = $l;
+                $lis[] = $this->getListItemTag($link, $li, $nav, $depth);
             }
         }
         if (count($lis) > 0) {
-            return '<ul class="rex-navi' . $depth . ' rex-navi-depth-' . $depth . ' rex-navi-has-' . count($lis) . '-elements">' . implode('', $lis) . '</ul>';
+            return $this->getListTag($lis, [
+                'class' => [
+                    'rex-navi'.$depth,
+                    'rex-navi-depth-'.$depth,
+                    'rex-navi-has-'.count($lis).'-elements',
+                ],
+            ], $depth);
         }
         return '';
     }
+
+    /**
+     * @param array $items
+     * @param array $attributes
+     * @return string
+     */
+    protected function getBreadcrumbListTag(array $items, array $attributes)
+    {
+         return '<ul'.rex_string::buildAttributes($attributes).'>'.PHP_EOL.implode('', $items).'</ul>'.PHP_EOL;
+    }
+
+    /**
+     * @param string $item
+     * @param array $attributes
+     * @return string
+     */
+    protected function getBreadcrumbListItemTag($item, array $attributes)
+    {
+        return '<li'.rex_string::buildAttributes($attributes).'>'.$item.'</li>'.PHP_EOL;
+    }
+
+    /**
+     * @param rex_category $category
+     * @param array $attributes
+     * @return string
+     */
+    protected function getBreadcrumbLinkTag(rex_category $category, array $attributes)
+    {
+        $category_name = htmlspecialchars($category->getName());
+        if (!isset($attributes['href'])) {
+            $attributes['href'] = $category->getUrl();
+        }
+
+        return '<a'.rex_string::buildAttributes($attributes).'>'.$category_name.'</a>';
+    }
+
+    /**
+     * @param array $items
+     * @param array $attributes
+     * @param int $depth
+     * @return string
+     */
+    protected function getListTag(array $items, array $attributes, $depth)
+    {
+        return '<ul'.rex_string::buildAttributes($attributes).'>'.PHP_EOL.implode('', $items).'</ul>'.PHP_EOL;
+    }
+
+    /**
+     * @param string $item
+     * @param array $attributes
+     * @param rex_category $category
+     * @param int $depth
+     * @return string
+     */
+    protected function getListItemTag($item, array $attributes, rex_category $category, $depth)
+    {
+        return '<li'.rex_string::buildAttributes($attributes).'>'.$item.'</li>'.PHP_EOL;
+    }
+
+    /**
+     * @param rex_category $category
+     * @param array $attributes
+     * @param int $depth
+     * @return string
+     */
+    protected function getLinkTag(rex_category $category, array $attributes, $depth)
+    {
+        $category_name = htmlspecialchars($category->getName());
```

Comment:
Das beißt sich mit https://github.com/redaxo/redaxo/pull/3092, /cc @alexplusde.
Da jetzt der Link-content hier in der Methode bestimmt wird, kann er nicht mehr über die Callbacks laufen.

Reicht uns die neue Variante aus diesem PR?
Oder übergeben wir den Link-content von außen an diese Methode?

Reply:
Das kann ich so nicht beurteilen, wie wäre denn die Variante hier?

Reply:
In diesem PR laufen alle Markup-Änderungen über Subklassen und dem Überschreiben der Methoden.
Wobei die erweiterte Klasse auch eine anonyme Klasse sein kann.

Den Linktext könnte man damit so ändern:

```php
$nav = new class extends rex_navigation {
    protected function getLinkTag(rex_category $category, array $attributes, $depth)
    {
        $content = $category->getName();
        $content = /* change content */;

        return '<a'.rex_string::buildAttributes($attributes).'>'.rex_escape($content).'</a>';
    }
};

// ab hier normal weiter mit $nav
```

Reply:
Zum Vergleich wie es mit deinem PR ausschaut:

```php
$nav = rex_navigation::factory();
$nav->addCallback(function (rex_category $category, $depth, &$li, &$a, &$a_content) {
    $a_content = $category->getName();
    $a_content = /* change content */;
});

Reply:
Es müssten immer extra Klassen erzeugt werden für verschiedene Navigationen, während ein Callback immer mit der Original-Klasse arbeitet. Das finde ich unpraktisch. Aber wenn mein PR das verhindert, muss er wohl weichen.

Reply:
rex_navigation eine Factory-Klasse, daher wäre der Weg meinem Verständnis nach:
```
// Neue Klasse erstellen
class my_navigation extends rex_navigation {
    ....
}

// Klasse anmelden
rex_navigation::setFactoryClass(may_navigation::class);

// Klasse nutzen
$nav = rex_navigation::factory();
echo $nav->show();
```
oder?

Reply:
Es gibt eigentlich drei Wege. 

Deiner ist der erste. Dabei wird dann an allen Stellen, wo man `rex_navigation::factory()` verwendet, die eigene Erweiterung genutzt.

Die zweite Variante ist, falls man nicht global überschreiben will, die Klasse genauso zu erstellen wie bei dir, aber dann nicht `setFactoryClass` aufrufen, und stattdessen bei der Nutzung direkt die eigene Klasse zu verwenden: 

```php
$nav = my_navigation::factory();
$nav->show();
```

Und die dritte oben von mir angedachte ist, direkt an Ort und Stelle die Klasse per anonymer Klasse zu erweitern:

```php
$nav = new class extends rex_navigation {
    public function __construct() {}

    protected function getLinkTag(rex_category $category, array $attributes, $depth)
    {
        // ...
    }
};

$nav->show();
```

Wobei mir da eben aufgefallen ist, dass man da den Constructor public machen muss (siehe Code). Falls wir den PR also aufnehmen, und die Variante aktiv bewerben wollten, sollte man überlegen, den Konstruktur in rex_navigation eventuell doch public zu machen (auch wenn es für die factory eigentlich blöd ist), damit man sich die Zeilen sparen könnte.

Reply:
Ah, OK. Ich überlege inzwischen, ob es nicht einfacher wäre, die ausgelagerten Methoden als Closure anzulegen, dann könnte man sich das Erstellen einer neuen Klasse sparen und das Überschreiben analog der Callbacks vornehmen. Das würde vielleicht die Komplexität reduzieren?

Reply:
Überlegt habe ich das auch, bin mir aber nicht sicher. Es würde halt die Komplexität innerhalb der Klasse erhöhen. Für viele Nutzer wäre es vermutlich aber in der Tat einfacher damit, da man keine separate Datei anlegen muss und viele mit der Alternative dazu, anonyme Klassen, noch nicht vertraut sind.

Bin so ein bisschen am verzweifeln, da ich mich selbst nicht entscheiden kann, was ich bevorzuge. Gar nicht aufnehmen, in der aktuellen Variante aufnehmen, oder in einer Callback-Variante aufnehmen.

Reply:
Nach einer Nach drüber schlafen: Ich würde bei der jetzigen Variante bleiben.

Aber dann noch mal zu meiner Ursprungsfrage hier im Thread, hast du da eine Meinung zu @DanielWeitenauer?
Also machen wir Alex PR rückgängig, oder ändern wir es hier so, dass an getLinkTag auch noch der Content (Kategoriename) geliefert wird, der eventuell über die Callback aus Alex PR abgeändert wurde?
Ich bin mir mal wieder unsicher.

Reply:
Da es ja bereits schon drin ist, und evtl. von einigen genutzt wird, ist wohl am sinnvollsten, den Namen zu übernehmen.

Reply:
Es ist erst für die neue 5.9 drin.
Aber bin ich mit einverstanden, wenn wir das so machen.

Reply:
Habe es eingebaut.